ark/phenotyping/preprocess.py

```diff
@@ -76,16 +78,16 @@ def create_pixel_matrix(img_xr, seg_labels, fovs=None, channels=None, blur_facto
 
         # turn into frequency
         pixel_mat.loc[:, channels] = pixel_mat.loc[:, channels].div(
-            pixel_mat.sum(axis=1), axis=0)
-
-        # 99.9% normalization
-        pixel_mat.loc[:, channels] = pixel_mat.loc[:, channels].div(
-            pixel_mat.quantile(q=0.999, axis=1), axis=0)
+            pixel_mat.loc[:, channels].sum(axis=1), axis=0)
 
         # assign to flowsom_data if not already assigned, otherwise concatenates
         if flowsom_data is None:
             flowsom_data = pixel_mat
         else:
             flowsom_data = pd.concat([flowsom_data, pixel_mat])
 
+    # 99.9% normalization
+    flowsom_data.loc[:, channels] = flowsom_data.loc[:, channels].div(
```

Comment:
@cliu72, do you do 99.9% normalization before or after converting to frequency?

Reply:
@ngreenwald looks like 99.9% normalization happens after converting to frequency.

Reply:
Candace, was there a reason you did it this way? I think it makes more sense to 99.9% threshold first and then convert to frequencies

Reply:
I did the frequency normalization first because it has the natural interpretation of calculating the normalized expression per pixel. Then the point of the 99.9% normalization is so that no single marker dominates the FlowSOM clustering. If we do 99.9% first, then the expression values for each pixel depend on the values of all the other pixels (which probably isn't an issue, but just has a less intuitive interpretation in my mind). Why do you think it makes more sense to apply 99.9% first? Since both of them are squishing the values to 0-1ish range, I don't know that it would make a big difference anyways.

Reply:
I can try it and see if it makes a difference.

Reply:
That makes sense. My thinking with 99.9% normalization is that the point is to find outliers within each channel. So first finding channel outliers, resetting them, and then using those corrected values to calculate the normalized pixel values. 

That way, each row will sum to 1 when being passed to the SOM.

Given the low dynamic range of the pixels, my suspicion is that this actually doesn't make a difference. But if you had a pixel that was super, super bright (i.e. well above the 99.9% threshold), if you row normalize and then 99.9% threshold, you're going to take a column that represents 99.99% of the data and modify it to be 99.9% of the data. 

If you did it the other way around, I think it would help to further reduce. But once again, given that most pixels are within 1-100, this likely won't have a big impact

Reply:
@ngreenwald so to clarify, should `preprocess.py` now run the 99.9% normalization before the frequency conversion?

Reply:
Hm I'm not totally sure I understand your example. Perhaps I'm misunderstanding what a 99.9% normalization does. I interpret it not as resetting outliers, but as getting all the channels in the same dynamic range. Even if I have a pixel that is super bright for marker A (above 99.9% threshold), even after a 99.9% normalization, that pixel would still be an "outlier" (it would be way above 1). The 99.9% normalization depends on the 99.9th percentile for marker A across all the pixels. Why would it modify it from 99.99% to be 99.9% of the data?

Reply:
@alex-l-kong I would hold off for now. If we want to change this step, I want to first try it on my data and see if it really is making a difference.

Reply:
Agreed, lets not change anything yet

Reply:
Okay, I ran the current processing pipeline on Erin's TB data that you sent Alex. Here's what I get for the 99.9th percentiles 
![image](https://user-images.githubusercontent.com/13770365/100686493-4d68e200-3333-11eb-8389-0abbf463258e.png)


Because we're row normalizing before percentile normalizing, many of the brightest markers have at least 0.1% of the pixels where they are the only signal. This means that the 99.9th percentile is 1. Therefore, dividing by the 99.9th percentile doesn't do anything for those markers.

It looks to me like the end result of this transformation is basically the increase the mean expression value of the dim markers. Perhaps z scoring/mean centering the raw channel data before processing would have a similar effect?

Reply:
In my experience, z-scoring isn't great because some of the lowly expressed markers have a mean of 0, so it's impossible for those z-scores to be below 0. The fact that some markers can have negative z-scores while others can't, seems to drive some weird clustering. If you think z scoring/mean centering will have a similar effect, is there a reason you want to do it instead of what we're doing now?

Reply:
I guess my main concern is that the current implementation of the 99.9 percent normalization isn't doing what we expect it to do, which would be to adjust the specific rows dominated by one marker. Instead, it's just multiplying all of the lowly expressed markers by a constant factor. It seems to be empirically working to resolve the shadow clusters, so I'm fine with keeping it. But if if we could figure out why it's useful, there might be a more accurate/better way to achieve the same result.

Reply:
@alex-l-kong we can skip a test for this part and leave it as is.

Reply:
Yeah I see what you mean. I guess in my mind, the 99.9 percent normalization isn't so much for adjusting specific rows dominated by one marker, but for adjusting the range of all the markers so they're approximately the same range. So for instance, if like Foxp3 is always 0.1 (or something super low), bring those values up so that we can pull out a Foxp3 cluster.

Reply:
For sure, sounds good!